graphql_compiler/schema_transformation/split_query.py

```diff
@@ -34,6 +39,340 @@ def __init__(self, query_ast):
+        if directives_from_existing_field is not None:
+            new_field.directives.extend(directives_from_existing_field)
+
+    # Transfer directives from edge
+    if directives_from_edge is not None:
+        for directive in directives_from_edge:
+            if directive.name.value == OutputDirective.name:  # output illegal on vertex field
+                raise GraphQLValidationError(
+                    u'Directive "{}" is not allowed on a vertex field, as @output directives '
+                    u'can only exist on property fields.'.format(directive)
+                )
+            elif directive.name.value == OptionalDirective.name:
+                if try_get_ast_by_name_and_type(
+                    new_field.directives, OptionalDirective.name, Directive
+                ) is None:
+                    # New optional directive
+                    new_field.directives.append(directive)
+            elif directive.name.value == FilterDirective.name:
+                new_field.directives.append(directive)
+            else:
+                raise AssertionError(
+                    u'Unreachable code reached. Directive "{}" is of an unsupported type, and '
+                    u'was not caught in a prior validation step.'.format(directive)
+                )
+
+    return new_field
+
+
+def _get_child_type_and_selections(ast, type_info):
+    """Get the type (root field) and selection set of the input AST.
+
+    For instance, if the root of the AST is a vertex field that goes to type Animal, the
+    returned type name will be Animal.
+
+    If the input AST is a type coercion, the root field name will be the coerced type rather
+    than the union or interface type, and the selection set will contain actual fields
+    rather than a single inline fragment.
+
+    Args:
+        ast: type Node, the AST that, if split off into its own Document, would have the the
+             type (named the same as the root vertex field) and root selections that this
+             function outputs
+        type_info: TypeInfo, used to check for the type that fields lead to
+
+    Returns:
+        Tuple[str, List[Union[Field, InlineFragment]]], name and selections of the root
+        vertex field of the input AST if it were made into its own separate Document
+    """
+    child_type = type_info.get_type()  # GraphQLType
+    if child_type is None:
+        raise SchemaStructureError(
+            u'The provided merged schema descriptor may be invalid, as the type '
+            u'corresponding to the field "{}" under type "{}" cannot be '
+            u'found.'.format(ast.name.value, type_info.get_parent_type())
+        )
+    child_type_name = strip_non_null_and_list_from_type(child_type).name
+
+    child_selection_set = ast.selection_set
+    # Adjust for type coercion
+    if (
+        child_selection_set is not None and
+        len(child_selection_set.selections) == 1 and
```

Comment:
What happens if someone passes an illegal query like:

```
{
    Foo {
        out_Foo_Bar {
            name
            ... on Baz {
                name
            }
        }
    }
}
```
or
```
{
    Foo {
        out_Foo_Bar {
            ... on Baz {
                name
            }
            ... on Quux {
                name
            }
        }
    }
}
```

Is this caught by any validation prior to this step?

Reply:
I'll fix this by adding in validation that inline fragments are alone in their scope.